oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,402 @@
+.. _edx-platform: https://github.com/edx/edx-platform
+
+Rationale
+=========
+
+The practices outlined here help prevent the following problems that we have
+encountered in the past:
+
+* A new deployment of an Open edX release fails because an unpinned indirect
+  dependency recently released a backwards-incompatible version.
+* Tests unrelated to a new code change fail, because an unpinned dependency
+  was upgraded to a backwards-incompatible version.  This can be difficult
+  to diagnose because the upgrade doesn't appear in the diff of pending
+  changes.
+* Tests have been running against a particular set of pinned versions for
+  years, but we now need to upgrade one (like Django) which requires also
+  upgrading several of the other dependencies.  This can force dealing with
+  a few years' worth of backwards-incompatible changes in multiple packages
+  all at once, whereas dealing with them one at a time every few months in
+  smaller pull requests would have been more manageable.
+* We have a different version of a dependency installed than we expect,
+  because the constraints imposed on pip for choosing a version vary between
+  different requirements files and we install them one file at a time.
+* We keep using years-old package versions despite the availability of newer
+  versions with accumulated bug fixes and performance improvements.
+* We install in production environments packages which are only needed for
+  testing, because we didn't make a clean distinction between the dependencies
+  for different usage contexts.  This slows down deployments.
+* We try to exhaustively pin all indirect dependencies manually, but miss some
+  (especially when a seemingly innocuous upgrade adds some new dependencies).
+* We keep installing a package long after we stopped using it, because nobody
+  remembers why it was added to the requirements file (especially true for
+  indirect dependencies that were later dropped as requirements of the package
+  we use directly).
+* We install an exhaustive set of testing dependencies in Travis, even though
+  we really only need it to run tox and codecov; the rest of the testing
+  dependencies are installed in a separate virtualenv created by tox, which
+  should have a separate requirements file.
+* An attempt to pin dependencies in setup.py (or parse its dependencies
+  automatically from a requirements file) forces us to change that package
+  before we can upgrade one of those dependencies in another repository
+  using that package.
+* We add a dependency without realizing that it requires multiple additional
+  indirect dependencies; we may have chosen an alternative if that had been
+  apparent.
+
+Reference Implementation
+========================
+
+Many of the Open edX repositories have already begun to comply with the
+recommendations outlined here.  In particular, repositories generated using
+`cookiecutter-django-app`_ should be configured correctly from the outset.
```

Comment:
very minor nit. Might be clearer to say "should already be configured correctly..."